pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,844 @@
+		logger.Info("not applying syncset since the observed generation of the syncset does not match the generation of the syncset")
+		syncStatus.Result = hiveintv1alpha1.FailureSyncSetResult
+		syncStatus.FailureMessage = "Current generation has not been observed yet"
+		return syncStatus, false
+	}
+	for _, cond := range syncSet.GetStatus().Conditions {
+		if cond.Type == hivev1.SyncSetInvalidResourceCondition &&
+			cond.Status == corev1.ConditionTrue {
+			logger.Info("not applying syncset as it contains an invalid resource")
+			syncStatus.Result = hiveintv1alpha1.FailureSyncSetResult
+			syncStatus.FailureMessage = "Invalid resource"
+			return syncStatus, false
+		}
+	}
+	applyFn := resourceHelper.Apply
+	applyFnMetricsLabel := labelApply
+	switch syncSet.GetSpec().ApplyBehavior {
+	case hivev1.CreateOrUpdateSyncSetApplyBehavior:
+		applyFn = resourceHelper.CreateOrUpdate
+		applyFnMetricsLabel = labelCreateOrUpdate
+	case hivev1.CreateOnlySyncSetApplyBehavior:
+		applyFn = resourceHelper.Create
+		applyFnMetricsLabel = labelCreateOnly
+	}
+	// Apply Resources
+	for i, resource := range syncSet.GetSpec().Resources {
+		identification := syncSet.GetStatus().Resources[i]
+		logger := logger.WithField("resourceNamespace", identification.Namespace).
+			WithField("resourceName", identification.Name).
+			WithField("resourceAPIVersion", identification.APIVersion).
+			WithField("resourceKind", identification.Kind)
+		logger.Debug("applying resource")
+		u := &unstructured.Unstructured{}
+		if err := yaml.Unmarshal(resource.Raw, u); err != nil {
+			logger.WithError(err).Warn("error decoding unstructured object")
+			syncStatus.Result = hiveintv1alpha1.FailureSyncSetResult
+			syncStatus.FailureMessage = fmt.Sprintf("Failed to decode resource %d: %v", i, err)
+			return syncStatus, false
+		}
+		if err := applyResource(u, applyFnMetricsLabel, applyFn, logger); err != nil {
+			syncStatus.Result = hiveintv1alpha1.FailureSyncSetResult
+			syncStatus.FailureMessage = fmt.Sprintf("Failed to apply resource %d: %v", i, err)
+			return syncStatus, true
+		}
+		if syncSet.GetSpec().ResourceApplyMode == hivev1.SyncResourceApplyMode {
+			syncStatus.ResourcesToDelete = append(syncStatus.ResourcesToDelete, identification)
+		}
+	}
+	// Apply Secrets
+	for i, secretMapping := range syncSet.GetSpec().Secrets {
+		logger := logger.WithField("secretNamespace", secretMapping.TargetRef.Namespace).
+			WithField("secretName", secretMapping.TargetRef.Name)
+		logger.Debug("applying secret")
+		secret := &corev1.Secret{}
+		if err := r.Get(context.Background(), types.NamespacedName{Namespace: syncSet.AsMetaObject().GetNamespace(), Name: secretMapping.SourceRef.Name}, secret); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "cannot read secret")
+			syncStatus.Result = hiveintv1alpha1.FailureSyncSetResult
+			syncStatus.FailureMessage = fmt.Sprintf("Failed to read secret %d: %v", i, err)
```

Comment:
the name/namespace of the source secret may be easier to parse than the index

Reply:
(unless that makes the status too big and then just ignore :) )

Reply:
I agree. This should use the name.

This also make me see that there is a bug in the namespace used for the source secret, too. I was going to say that the namespace is not necessary in the failure message since the secret must be in the same namespace as the syncset. But, for SelectorSyncSets, the namespace must be specified. The Get is incorrectly always using the namespace of the syncset.

Reply:
Upon more consideration, I would like to keep using the index here. That keeps it consistent with the failure messages for resources and patches. Also, it is possible for there to be multiple secret mappings specified that use the same source secret.

Reply:
👍